src/packages/utils/src/things/json-rpc/json-rpc-data.ts

```diff
@@ -1,57 +1,107 @@
-  }
-}
+type JsonRpcDataInputArg = string | Buffer;
 
 export class Data extends BaseJsonRpcType {
-
-  constructor(value: string | Buffer, private _byteLength?: number) {
+  constructor(value: JsonRpcDataInputArg, private _byteLength?: number) {
     super(value);
+
     if (typeof value === "bigint") {
       throw new Error(`Cannot create a ${typeof value} as a Data`);
     }
     if (_byteLength !== undefined) {
-      validateByteLength(_byteLength);
+      Data.validateByteLength(_byteLength);
     }
   }
-  public toString(byteLength?: number): string {
-    if (byteLength === undefined) {
-      byteLength = this._byteLength;
+
+  public toString(byteLength?: number): string | null {
+    Data.validateByteLength(byteLength);
+    const length = byteLength || this._byteLength;
+
+    if (this.bufferValue == null) {
+      return <null>this.bufferValue;
+    }
+
+    if (length === undefined) {
+      return super.toString();
+    }
+
+    return `0x${Data.padString(this.bufferValue.toString("hex"), length)}`;
+  }
+
+  public toBuffer(byteLength?: number): Buffer {
+    Data.validateByteLength(byteLength);
+
+    const length = byteLength || this._byteLength;
+
+    if (this.bufferValue == null) {
+      return BUFFER_EMPTY;
+    }
+
+    if (length === undefined || length === this.bufferValue.length) {
+      return this.bufferValue;
+    }
+
+    return Data.padBuffer(this.bufferValue, length);
+  }
+
+  public static from(value: JsonRpcDataInputArg, byteLength?: number) {
+    if (value instanceof Data) {
+      return value;
     }
-    if (byteLength === undefined && strCache.has(this)) {
-      return strCache.get(this) as string;
-    } else {
-      let str = toStrings.get(this)() as string;
-      let length = str.length;
-
-      if (length % 2 === 1) {
-        length++;
-        str = `0${str}`;
-      }
-
-      if (byteLength !== undefined) {
-        validateByteLength(byteLength);
-        const strLength = byteLength * 2;
-        const padBy = strLength - length;
-        if (padBy < 0) {
-          // if our hex-encoded data is longer than it should be, truncate it:
-          str = str.slice(0, strLength);
-        } else if (padBy > 0) {
-          // if our hex-encoded data is shorter than it should be, pad it:
-          str = "0".repeat(padBy) + str;
-        }
-      }
-      return `0x${str}`;
-    }
-  }
-  public static from<T extends string | Buffer = string | Buffer>(
-    value: T,
-    byteLength?: number
-  ) {
     return new Data(value, byteLength);
   }
+
+  private static validateByteLength(byteLength?: number) {
+    if (byteLength !== undefined && (typeof byteLength !== "number" || byteLength < 0 || !isFinite(byteLength))) {
+      throw new Error(`byteLength must be a number greater than or equal to 0, provided: ${byteLength}`);
+    }
+  }
+  
+  private static padString(value: string, byteLength: number | undefined) {
```

Comment:
`padString` _feels_ like a weird name to me. I don't know if it's a formal definition, but to me padding is additive, but this function can go either way. Maybe `padOrSliceString`? Or `correctStringLength`? ¯\_(ツ)_/¯

Reply:
Agreed - I changed to "bufferToFixedByteLength" and "stringToFixedByteLength". (I originally has stringToFixedLength, but it was strange that we then passed in a _byte_length rather than char length. I made the string and buffer signatures match, even though it made sense to pass byte length to bufferToFixedLength).